ttnn/cpp/ttnn/operations/data_movement/pad/pad.hpp

```diff
@@ -0,0 +1,144 @@
+        uint8_t queue_id,
+        const ttnn::Tensor& input_tensor,
+        std::vector<std::pair<uint32_t, uint32_t>> padding, //intentionally not const&
+        const float value,
+        const bool use_multicore,
+        const std::optional<MemoryConfig>& memory_config_arg) {
+
+        const int original_rank = input_tensor.get_shape().rank();
+        if(int diff = original_rank - padding.size(); diff != 0) {
+            TT_FATAL(diff > 0, "ttnn.pad: padding len can't be larger than input tensor rank");
+
+            padding.insert(padding.begin(), diff, {0, 0});
+        }
+
+        TT_FATAL(
+            padding.size() == original_rank,
+            "ttnn.pad: padding must be the same length as the input tensor rank");
+        TT_FATAL(
+            input_tensor.get_layout() != ttnn::ROW_MAJOR_LAYOUT,
+            "ttnn.pad: row-major tensors have to use fallback because the kernel currently causes a PCC error");
+
+        // Unsqueeze Tensor to 4D if it is not already
+        ttnn::Tensor input_tensor_4D = ttnn::unsqueeze_to_4D(input_tensor);
+        padding.insert(padding.begin(), 4 - original_rank, {0, 0});
+        auto input_shape_with_tile_padding = input_tensor_4D.get_shape().with_tile_padding();
+        std::vector<uint32_t> output_padded_shape(padding.size());
+        for(size_t i = 0; i < padding.size(); i++) {
+            output_padded_shape[i] = input_shape_with_tile_padding[i] + padding[i].second;
+        }
+
+        // Due to the strangeness of tt::tt_metal::pad, we need to split front and back pad
+        // Front will be passed separately. And pad_back is retrieved -> output_padded_shape - pad_front
+        auto memory_config = memory_config_arg.value_or(input_tensor.memory_config());
+        auto pad_front = padding | std::views::transform([](const auto& p) { return p.first; });
+        auto pad_back = padding | std::views::transform([](const auto& p) { return p.second; });
+
+        const bool front_padding_is_zero = std::accumulate(pad_front.begin(), pad_front.end(), 0) == 0;
+        TT_FATAL(
+            front_padding_is_zero,
+            "ttnn.pad: on device padding does not support front padding");
+
+        const int target_height = output_padded_shape[output_padded_shape.size() - 2];
+        const int target_width = output_padded_shape[output_padded_shape.size() - 1];
+        TT_FATAL(
+            target_height % ttnn::TILE_SIZE == 0 || target_width % ttnn::TILE_SIZE == 0,
+            "ttnn.pad: for tiled tensors padding end must be a multiple of the tile size on height and width for a "
+            "tensor in tile layout");
+
+        // Performing actual padding
+        std::vector<uint32_t> pad_front_vec(pad_front.begin(), pad_front.end());
+        auto output_tensor = operation::run(
+            tt::tt_metal::Pad{
+                .output_tensor_shape=tt::tt_metal::Shape(output_padded_shape),
+                .input_tensor_start=tt::tt_metal::Shape(pad_front_vec),
```

Comment:
Switch to `ttnn`?

Reply:
that's coming in a future PR, this PR migrated the host portion but its still calling the device side code from the tt eager folder.